transport/session/src/balancer/controller.rs

```diff
@@ -0,0 +1,299 @@
+            &["session_id"]
+    ).unwrap();
+    static ref METRIC_CONTROL_OUTPUT: hopr_metrics::metrics::MultiGauge =
+        hopr_metrics::metrics::MultiGauge::new(
+            "hopr_surb_balancer_control_output",
+            "hopr_surb_balancer_control_output",
+            &["session_id"]
+    ).unwrap();
+    static ref METRIC_SURBS_CONSUMED: hopr_metrics::metrics::MultiCounter =
+        hopr_metrics::metrics::MultiCounter::new(
+            "hopr_surb_balancer_surbs_consumed",
+            "Estimations of the number of SURBs consumed by the counterparty",
+            &["session_id"]
+    ).unwrap();
+    static ref METRIC_SURBS_PRODUCED: hopr_metrics::metrics::MultiCounter =
+        hopr_metrics::metrics::MultiCounter::new(
+            "hopr_surb_balancer_surbs_produced",
+            "Estimations of the number of SURBs produced for the counterparty",
+            &["session_id"]
+    ).unwrap();
+}
+
+/// Configuration for the [`SurbBalancer`].
+#[derive(Clone, Copy, Debug, PartialEq, Eq, smart_default::SmartDefault)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct SurbBalancerConfig {
+    /// The desired number of SURBs to be always kept as a buffer at the Session counterparty.
+    ///
+    /// The [`SurbBalancer`] will try to maintain approximately this number of SURBs
+    /// at the counterparty at all times, by regulating the [flow of non-organic SURBs](SurbFlowController).
+    #[default(5_000)]
+    pub target_surb_buffer_size: u64,
+    /// Maximum outflow of non-organic surbs.
+    ///
+    /// The default is 2500 (which is 1250 packets/second currently)
+    #[default(2_500)]
+    pub max_surbs_per_sec: u64,
+}
+
+/// Runs a continuous process that attempts to [evaluate](SurbFlowEstimator) and
+/// [regulate](SurbFlowController) the flow of non-organic SURBs to the Session counterparty,
+/// to keep the number of SURBs at the counterparty at a certain level.
+///
+/// Internally, the Balancer uses a PID (Proportional Integral Derivative) controller to
+/// control the rate of SURBs sent to the counterparty
+/// each time the [`update`](SurbBalancer::update) method is called:
+///
+/// 1. The size of the SURB buffer at the counterparty is estimated using [`SurbFlowEstimator`].
+/// 2. Error against a set-point given in [`SurbBalancerConfig`] is evaluated in the PID controller.
+/// 3. The PID controller applies a new SURB flow rate value using the [`SurbFlowController`].
+pub struct SurbBalancer<I, O, F, S> {
+    session_id: S,
+    pid: Pid<f64>,
+    surb_production_estimator: O,
+    surb_consumption_estimator: I,
+    controller: F,
+    cfg: SurbBalancerConfig,
+    current_target_buffer: u64,
+    last_surbs_delivered: u64,
+    last_surbs_used: u64,
+    last_update: std::time::Instant,
+    was_below_target: bool,
+}
+
+// Default coefficients for the PID controller
+// This might be tweaked in the future.
+const DEFAULT_P_GAIN: f64 = 0.6;
```

Comment:
Should this be configurable, maybe hidden, as well? That way we could actually conceive automatic tests to find good settings.

Reply:
I can make it configurable via env vars.